File: app/src/main/java/com/zoneanchor/SystemEventReceiver.java
package com.zoneanchor;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

public final class SystemEventReceiver extends BroadcastReceiver {
    @Override
    public void onReceive(Context context, Intent intent) {
        for (ZonedAlarm alarm : AlarmStore.loadAll(context)) {
            if (alarm.enabled) {
                AlarmScheduler.scheduleNext(context, alarm);
            }
        }
    }
}

File: app/src/main/java/com/zoneanchor/alarm/SystemEventReceiver.kt
package com.zoneanchor.alarm

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import com.zoneanchor.ZoneAnchorApp
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

open class SystemEventReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val pending = goAsync()
        scope.launch {
            try {
                (context.applicationContext as ZoneAnchorApp).alarmScheduler.rescheduleEnabledAlarms()
            } finally {
                pending.finish()
            }
        }
    }

    companion object {
        private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    }
}
